_posts/2019-10-16-deployment-updates.md

```diff
@@ -0,0 +1,214 @@
+JupyterLab extension, and so on are more consistently available.  Also, we've
+been able to unify development and reduce our maintenance burden considerably.
+
+
+Dask-CloudProvider
+------------------
+
+For cloud deployments we generally recommend using a hosted Kubernetes or Yarn
+service, and then using Dask-Kubernetes or Dask-Yarn on top of these.
+
+However in some institutions these hosted services aren't yet accessible, and
+it's more convenient to use APIs that are more native to the particular cloud.
+The new package [Dask Cloudprovider](https://cloudprovider.dask.org) handles
+this today for Amazon's ECS API, which has been around for a long while and is
+more universally accepted.
+
+
+```python
+from dask_cloudprovider import ECSCluster
+cluster = ECSCluster(cluster_arn="arn:aws:ecs:<region>:<acctid>:cluster/<clustername>")
+
+from dask_cloudprovider import FargateCluster
+cluster = FargateCluster()
+```
+
+Dask-Gateway
+------------
+
+In some cases users may not have access to the cluster manager.  For example
+the institution may not give all of their data science users access to the Yarn
+or Kubernetes cluster.  In this the [Dask-Gateway](https://gateway.dask.org)
+project may be useful.
+It can launch and manage Dask jobs,
+and provide a proxy connection to these jobs if necessary.
+It is typically deployed with elevated permissions but managed directly by IT,
+giving them a point of greater control.
+
+<img src="https://gateway.dask.org/_images/architecture.svg" width="50%">
+
+
+GPUs and Dask-CUDA
+------------------
+
+While using Dask with multi-GPU deployments the [NVIDIA
+Rapids](https://rapids.ai) has needed the ability to specify increasingly
+complex setups of Dask workers.  They recommend the following deployment
+strategy:
+
+1.  One Dask-worker per GPU on a machine
+2.  Specify the `CUDA_VISIBLE_DEVICES` environment variable to pin that worker
+    to that GPU
+3.  If your machine has multiple network interfaces then choose the network interface closest to that GPU
+4.  If your machine has multiple CPUs then set thread affinities to use the closest CPU
+5.  ... and more
+
+For this reason we wanted to specify these configurations in code, like the
+following:
+
+```python
+specification = {
+    "worker-0": {
+        "cls": dask.distributed.Nanny,
+        "options": {"nthreads": 1, "env": {"CUDA_VISIBLE_DEVICES": "0,1,2,3"}, interface="ib0"},
+    },
+    "worker-1": {
+        "cls": dask.distributed.Nanny,
+        "options": {"nthreads": 1, "env": {"CUDA_VISIBLE_DEVICES": "1,2,3,0"}, interface="ib0"},
+    },
+    "worker-2": {
+        "cls": dask.distributed.Nanny,
+        "options": {"nthreads": 1, "env": {"CUDA_VISIBLE_DEVICES": "2,3,0,1"}, interface="ib1"},
+    },
+    "worker-2": {
+        "cls": dask.distributed.Nanny,
+        "options": {"nthreads": 1, "env": {"CUDA_VISIBLE_DEVICES": "3,0,1,2"}, interface="ib1"},
+    },
+}
+```
```

Comment:
I've added the `worker_spec` function which acts as a "recipe" to simplify that process in dask-cuda. Perhaps it's worth mentioning that too or instead of this long spec?